tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -984,6 +998,275 @@ TEST_P(DistributedTransformerTest, Forward) {
+  TensorView* mha_mask = makeContigTensor(2, DataType::Bool);
+  TensorView* mlp_mask = makeContigTensor(2, DataType::Bool);
+  TensorView* mha_sdpa_out = makeConcreteTensor({D, B, H / D, S, E / H}, dtype);
+  TensorView* mha_sdpa_log_sumexp =
+      makeContigConcreteTensor({D, B, H / D, S}, DataType::Float);
+  TensorView* mha_sdpa_seed = makeSymbolicTensor({}, DataType::Int);
+  TensorView* mha_sdpa_offset = makeSymbolicTensor({}, DataType::Int);
+  TensorView* ln1_w = makeContigTensor(1);
+  TensorView* ln1_b = makeContigTensor(1);
+  TensorView* ln1_mean = makeConcreteTensor({256, 1});
+  TensorView* ln1_rstd = makeConcreteTensor({256, 1});
+  TensorView* ln0_w = makeContigTensor(1);
+  TensorView* ln0_b = makeContigTensor(1);
+  TensorView* ln0_mean = makeConcreteTensor({256, 1});
+  TensorView* ln0_rstd = makeConcreteTensor({256, 1});
+  TensorView* mha_linear1 = makeContigTensor(2);
+
+  fusion->addInput(x);
+  fusion->addInput(grad);
+  fusion->addInput(mha_w0);
+  fusion->addInput(mha_b0);
+  fusion->addInput(mha_w1);
+  fusion->addInput(mha_b1);
+  fusion->addInput(mlp_w0);
+  fusion->addInput(mlp_b0);
+  fusion->addInput(mlp_w1);
+  fusion->addInput(mlp_b1);
+  fusion->addInput(mlp_mask);
+  fusion->addInput(mha_mask);
+  fusion->addInput(mha_sdpa_out);
+  fusion->addInput(mha_sdpa_log_sumexp);
+  fusion->addInput(mha_sdpa_seed);
+  fusion->addInput(mha_sdpa_offset);
+  fusion->addInput(ln1_w);
+  fusion->addInput(ln1_b);
+  fusion->addInput(ln1_mean);
+  fusion->addInput(ln1_rstd);
+  fusion->addInput(ln0_w);
+  fusion->addInput(ln0_b);
+  fusion->addInput(ln0_mean);
+  fusion->addInput(ln0_rstd);
+  fusion->addInput(mha_linear1);
+
+  const auto D = mha_w0->axis(0)->extent()->value().as<int64_t>();
+  // Recomputation
+  auto ln_0 = layer_norm(x, norm_shape, ln0_w, ln0_b, eps);
+  auto mha_in = castOp(dtype, std::get<0>(ln_0));
+  auto qkv = mha_qkv(mha_in, mha_w0, mha_b0, mha_w1, mesh);
+  // The thunder trace recompute mha linear1, but this would result in 3
+  // AllReduces in the backwards pass.
+  if (mha_linear1 == nullptr) {
+    TensorView* sdpa_transpose = transpose(mha_sdpa_out, 2, 3);
+    TensorView* sdpa_reshape =
+        reshape(sdpa_transpose, {D, B, S, H / D, E / H}, {D, B * S, E / D});
+    TensorView* mha_local_matmul1 = matmul(sdpa_reshape, mha_w1);
+    TensorView* mha_matmul1 = sum(mha_local_matmul1, {0}); // allreduce
+    mha_linear1 = add(mha_matmul1, broadcast(mha_b1, {true, false}));
+  }
+  Val* scale = IrBuilder::create<Val>(1.0 / (1.0 - kDropoutProb));
+  // Use input mha mask to implement dropout
+  auto mha_out = mul(mha_linear1, mha_mask);
+  mha_out = mul(mha_out, scale);
+  auto resid_0 = add(x, mha_out);
+  auto ln_1 = layer_norm(resid_0, norm_shape, ln1_w, ln1_b, eps);
+  auto mlp_in = castOp(dtype, ln_1.output);
+  // Note: We only use linear0 and gelu outputs from the mlp forward pass.
+  auto mlp_tensors = mlp(mlp_in, mlp_w0, mlp_b0, mlp_w1, mlp_b1, mesh);
```

Comment:
I like how you made LinearBackwardsResult to encapsulate things into a struct with named fields. You may want to consider doing that for MLP results as well, so (1) the readers are clear about the semantics of `mlp_tensors[0]` and `mlp_tensors[1]` and (2) it's harder to make mistakes if we change `mlp`'s return values.

Reply:
(Of course, this can come in a separate PR)

Reply:
Yes, I was thinking about doing that for all the helpers. Probably for a later PR since this one is getting quite big!